snapcraft/internal/project_loader/_extensions/_utils.py

```diff
@@ -156,25 +158,58 @@ def _apply_extension(
     part_extension = extension.part_snippet
     parts = yaml_data["parts"]
     for part_name, part_definition in parts.items():
+
         for property_name, property_value in part_extension.items():
+            # part_definition_unordered[property_name] is a dict
             part_definition[property_name] = _apply_extension_property(
                 part_definition.get(property_name), property_value
             )
 
+        # Stores the extension's list of part_snippets in each part
+        parts[part_name] = part_definition
+
     # Finally, add any parts specified in the extension
     for part_name, part_definition in extension.parts.items():
         # If a extension part name clashes with a part that already exists, error.
         if part_name in parts:
             raise errors.ExtensionPartConflictError(extension_name, part_name)
 
+        # Stores the extension's list of parts in the parts section
         parts[part_name] = part_definition
 
 
 def _apply_extension_property(existing_property: Any, extension_property: Any):
     if existing_property:
-        # If the property is not scalar, merge them
+
         if isinstance(existing_property, list) and isinstance(extension_property, list):
-            return _merge_lists(existing_property, extension_property)
```

Comment:
Could you provide some additional context / examples on what's being done here?

While outside the scope of this PR, I _really_ think we should reconsider the approach for extensions.  Why not operate on the YAML directly (and eventually, being a Snap object)? There's a lot of complexity around what amounts to a handful trivial operations on a dictionary and a base validation check. :disappointed:

Reply:
I added clarity and tried to not be too wordy so please take a look and let me know if that makes sense.

In the long term, we're planning to use the freedesktop-sdk and the corresponding gnome runtimes (1-2 years). So for now, I think this method of applying the extension could hold us over until then, but I'm open to changing things sooner if we feel the need to do so.

Reply:
Thank you! I think I understand the problem now.  Everything is complicated because of the use of a list of dictionaries for `build-environment`, instead of a simple dict/object.

I expect the use of the list in the case of `build-environment` is to guarantee ordering.  The irony here of course is that each element in build-environment gets parsed into its own OrderedDict, but it's really effectively a combined OrderedDict.  If it were, then you could just use '.update()' to merge the dictionaries, e.g.:

```
>>> from collections import OrderedDict
>>> od = OrderedDict()
>>> od['a'] = 1
>>> od['b'] = 2
>>> od['c'] = 3
>>> od['d'] = 4
>>>
>>> od2 = OrderedDict()
>>> od2['e'] = 5
>>> od2['f'] = 6
>>> od2['a'] = 7
>>> od2['d'] = 8
>>>
>>> merged = od.copy()
>>> merged.update(od2)
>>> merged
OrderedDict([('a', 7), ('b', 2), ('c', 3), ('d', 8), ('e', 5), ('f', 6)])
```

Which I think satisfies what you're trying to do?


I pondered it for a bit and I would suggest something like:
https://github.com/cjp256/snapcraft/commit/576ccae204c5784e01b2907bc8698aed12ee8504

Namely, eliminate any assumptions that will come back to bite us later (special case handling only for build-environment rather than assuming what to do based on some incomplete type information and context available here).  And I have a preference towards using the OrderedDict update operation, because that's what the list really is :smile: 

But whatever you decide, can you add a few of unit tests to cover some scenarios regarding this merge functionality?  It's fairly complex, and deserves a bit of attention to ensure we're not going to introduce some regressions.